// @sudoku/stores/gameStore.js
/**
 * GameStore 
 * 工作原理：
 * 1. 内部持有 createGame() 实例
 * 2. 使用 Svelte writable store 暴露响应式状态
 * 3. 方法调用后更新 store，触发 UI 刷新
 * 4. 组件通过 $gameStore 获取整个状态对象
 */

import { writable, derived } from 'svelte/store';
import { createGame, createSudoku } from '../../../domain/index.js';
import { generateSudoku } from '@sudoku/sudoku';
import { decodeSencode } from '@sudoku/sencode';

const SUDOKU_SIZE = 9;
const BOX_SIZE = 3;

/**
 * 创建 GameStore - 返回合法的 Svelte store
 * 
 * @returns {Store} 符合 Svelte store 约定的对象
 */
function createGameStore() {
    // 内部持有的领域对象
    let game = null;

    // 初始状态
    const initialState = {
        grid: [
            [0,0,0,0,0,0,0,0,0],
            [0,0,0,0,0,0,0,0,0],
            [0,0,0,0,0,0,0,0,0],
            [0,0,0,0,0,0,0,0,0],
            [0,0,0,0,0,0,0,0,0],
            [0,0,0,0,0,0,0,0,0],
            [0,0,0,0,0,0,0,0,0],
            [0,0,0,0,0,0,0,0,0],
            [0,0,0,0,0,0,0,0,0],
        ],
        initialGrid: null,  // 题面（不可编辑）
        canUndo: false,
        canRedo: false,
        won: false,
        invalidCells: [],
    };

    // 主 store
    const { subscribe, set, update } = writable(initialState);

    // 计算无效单元格
    function computeInvalidCells(grid) {
        const invalid = [];
        const addInvalid = (x, y) => {
            const key = x + ',' + y;
            if (!invalid.includes(key)) invalid.push(key);
        };

        for (let y = 0; y < SUDOKU_SIZE; y++) {
            for (let x = 0; x < SUDOKU_SIZE; x++) {
                const value = grid[y][x];
                if (value) {
                    for (let i = 0; i < SUDOKU_SIZE; i++) {
                        if (i !== x && grid[y][i] === value) addInvalid(x, y);
                        if (i !== y && grid[i][x] === value) addInvalid(x, i);
                    }
                    const startY = Math.floor(y / BOX_SIZE) * BOX_SIZE;
                    const startX = Math.floor(x / BOX_SIZE) * BOX_SIZE;
                    for (let row = startY; row < startY + BOX_SIZE; row++) {
                        for (let col = startX; col < startX + BOX_SIZE; col++) {
                            if (row !== y && col !== x && grid[row][col] === value) {
                                addInvalid(col, row);
                            }
                        }
                    }
                }
            }
        }
        return invalid;
    }

    // 计算是否获胜
    function computeWon(grid, invalidCells) {
        for (let y = 0; y < SUDOKU_SIZE; y++) {
            for (let x = 0; x < SUDOKU_SIZE; x++) {
                if (grid[y][x] === 0) return false;
            }
        }
        return invalidCells.length === 0;
    }

    // 同步领域对象状态到 store（支持普通模式和探索模式）
    function syncFromDomain() {
        if (!game) return;
        
        const gameState = game.getGameState();
        const newGrid = gameState.sudoku.getGrid();
        
        update(state => {
            const invalidCells = computeInvalidCells(newGrid);
            return {
                ...state,
                grid: newGrid,
                canUndo: gameState.canUndo,
                canRedo: gameState.canRedo,
                invalidCells,
                won: computeWon(newGrid, invalidCells),
                isExploring: gameState.isExploring,
                isExploreFailed: gameState.isExploreFailed,
                exploreHistoryLength: gameState.exploreHistoryLength,
                exploreHistoryIndex: gameState.exploreHistoryIndex,
            };
        });
    }

    // 同步初始题面
    function syncInitialGrid() {
        if (!game) return;
        
        // 从初始 Sudoku 获取题面
        const initialSudoku = game.getInitialSudoku ? game.getInitialSudoku() : game.getSudoku();
        const initialGrid = initialSudoku.getGrid();
        
        update(state => ({
            ...state,
            initialGrid,
        }));
    }

    /**
     * 开始新游戏
     */
    function startNew(difficulty) {
        const puzzle = generateSudoku(difficulty);
        const sudoku = createSudoku(puzzle);
        game = createGame({ sudoku });
        
        syncInitialGrid();
        syncFromDomain();
    }

    /**
     * 从 sencode 加载
     */
    function startCustom(sencode) {
        const puzzle = decodeSencode(sencode);
        const sudoku = createSudoku(puzzle);
        game = createGame({ sudoku });
        
        syncInitialGrid();
        syncFromDomain();
    }

    /**
     * 用户猜测（支持普通模式/探索模式）
     */
    function guess(pos, value) {
        if (!game) return;
        
        // 检查是否可编辑（不是题面格子）
        const state = getState();
        if (state.initialGrid && state.initialGrid[pos.y][pos.x] !== 0) {
            return; // 不能修改题面
        }
        
        game.currentGuess({ row: pos.y, col: pos.x, value });
        syncFromDomain();
    }

    /**
     * 撤销（支持普通模式/探索模式）
     */
    function undo() {
        if (!game) return;
        if (game.isExploring()) {
            game.exploreUndo();
        } else {
            game.undo();
        }
        syncFromDomain();
    }

    /**
     * 重做
     */
    function redo() {
        if (!game) return;
        if (game.isExploring()) {
            game.exploreRedo();
        } else {
            game.redo();
        }
        syncFromDomain();
    }

    /**
     * 开始探索模式
     */
    function startExplore() {
        if (!game) return;
        game.startExplore();
        syncFromDomain();
    }

    /**
     * 提交探索结果
     */
    function commitExplore() {
        if (!game) return;
        game.commitExplore();
        syncFromDomain();
    }

    /**
     * 取消探索
     */
    function cancelExplore() {
        if (!game) return;
        game.cancelExplore();
        syncFromDomain();
    }

    /**
     * 获取提示（候选数）
     */
    function getHint(pos) {
        if (!game) return null;
        const sudoku = game.getCurrentSudoku();
        if (!sudoku) return null;
        
        if (pos) {
            return sudoku.getCandidates(pos.y, pos.x);
        } else {
            return sudoku.getNextHint();
        }
    }

    // 获取当前状态（内部使用）
    function getState() {
        let currentState;
        subscribe(s => currentState = s)();
        return currentState;
    }

    // 返回符合 Svelte store 约定的对象
    return {
        subscribe,
        startNew,
        startCustom,
        guess,
        undo,
        redo,
        startExplore,
        commitExplore,
        cancelExplore,
        getHint,
    };
}

// 导出 store
export const gameStore = createGameStore();
